Fix rule expiration

Darkren opened the pull request:
Partially fixes #147 	

 Changes:	
- Now rule activity gets updated automatically only for consume type of rules (to avoid unnecessary mutex acquiring for the second time in a row)
- In all other cases route groups tells router to update rule activity ONLY if the packet was propagated successfully. This way when transport goes down, rules will be garbage collected and the route will disappear after a timeout
- Routing table is now being initialized inside a router constructor. All outer calls to the table are proxied through the router instance
- Timed out rules garbage collecting routine is now being called from the router, GC loop is moved there
- If consume rule gets timed out, router fetches the corresponding route group, closes it and totally removes. This allows outer code to react properly when remote visor gets killed unexpectedly
- Some route ID errors were fixed. At a couple of places packets were composed with the wrong ID which worked well in the freshly started integration env but actually started failing after a new route got established next to the old one
- Skysocks client is refactored a bit. Now we detect if app conn is failed at `DefaultRouteKeepAlive / 2` time tops. Then app conn is considered broken, client gets closed and reinstantiated by redialing app conn and recreating route. Redialing happens with an exponential backoff
- If dialing visor gets restarted, new route is established instantly. If remove visor (or any visor in between) gets restarted, new route is established as soon as dialing visor sees that route is broken (when route times out and route group gets closed)
- Route timeout reduced to 30 seconds, keep-alive internal is bound to be half of it (15 seconds for now)

Darkren committed:
Move rule activity update to after the packet successful propagation

Darkren commented:
So, the simplest part is done, moving forward

Darkren committed:
Start fixing

Darkren committed:
Move a bit further

Darkren committed:
Update vendor

Darkren committed:
Move routing table initialization inside the router. Proxy all visor RPC calls to RT through router interface

Darkren committed:
Move rules GC routing from routing table to router

Darkren committed:
Merge branch 'milestone2' of github.com:SkycoinProject/skywire-mainnet into fix/rule-expiration

Darkren committed:
Complete rules GC loop, remove old unused code, refactor a bit

Darkren committed:
Fix non-pointer method call

Darkren committed:
Fix

Darkren committed:
Regenerate router mock

Darkren committed:
Fix `router` tests, remove test for routing table clenup due to the corresponding functionality removal

evanlinjin reviewed (commented):
Would `RuleLastAccessed` be a better term than `RuleActivity`?

evanlinjin commented on pkg/router/router.go:
```diff
@@ -634,3 +659,72 @@ func (r *router) IntroduceRules(rules routing.EdgeRules) error {
 }
+
+// RoutesCount returns count of the routes stored within the routing table.
+func (r *router) RoutesCount() int {
```
Would it be easier just to have a function: `(*router).Table() routing.Table` to reduce clutter?

Darkren replied:
@evanlinjin well, I just proxied all calls to the table through router, there's outer call from visor to get this count, so I left it as is to ensure compatibility

evanlinjin commented on pkg/routing/table.go:
```diff
@@ -40,43 +39,26 @@ type Table interface {
 	Count() int
+
+	// CollectGarbage checks all the stored rules, removes and returns ones that timed out.
+	CollectGarbage() []Rule
```
What's the reasoning behind returning `Rule` instead of `RouteDescriptor`?

Darkren replied:
@evanlinjin routing table stores separate rules, not descriptors. descriptors are part of a route group which stands a bit above the table, table itself is rather simple

evanlinjin reviewed (commented):
Good work so far :)

Darkren committed:
Update vendor

Darkren committed:
Add debug logs

Darkren committed:
Fix route group closing on timeout

Darkren committed:
Update vendor

Darkren committed:
Fix skychat conn removal, add A LOT OF debug logs

Darkren committed:
Merge branch 'milestone2' of github.com:SkycoinProject/skywire-mainnet into fix/rule-expiration

Darkren committed:
Add debug logs, remove last `hashicorp/yamux` reference

Darkren committed:
Disable yamux internal keep-alive for skysocks

Darkren committed:
Fix packet forwarding from route group after a new route establishment

Darkren committed:
Remove yamux internal keep-alive for skysocks client

Darkren commented:
Tests. All running in the proxy integration env (skysocks-client on C and skysocks on A):
1) curl, `kill -9` visor C, wait for route to timeout, start it again, curl again
2) curl, `kill -9` visor C, restart it after a couple of seconds, immediately curl as soon as it starts
3) repeat 1) for visor A
4) repeat 2) for visor A
5) repeat 1) for visor B
6) repeat 2) for visor B

- STCP:
- [x] 1)
- [x] 2)
- [x] 3)
This one won't work, since visor C is responsible for route establishing. And app actually quits as soon as the route times out. We should add a route re-establishment with the exponential backoff for such situations
- [x] 4)
Same as 3) above
- [x] 5)
- [x] 6)

- DMSG:
- [ ] 1)
Looks exactly the same as 2) described below
- [ ] 2)
When I do this, I get the following logs on the restarted visor C:
```
time="2020-02-25T08:38:34+03:00" level=warning msg="route finder: no route to destination" source="retrier.go:90"
```
Then after a couple of seconds:
```
failed to redial underlying connection (redial loop): settlement handshake failed: context deadline exceeded
Failed to update statuses error="Post http://localhost:9091/statuses: context deadline exceeded"
```
Then transport gets `UP` status but logs show:
```
 WARN [tp:0348c9]: failed to read packet: EOF
```
And this line just keeps spamming

- [ ] 3)
- [ ] 4)
- [ ] 5)
- [ ] 6)

Darkren commented:
2), 3) and 4) for STCP show strange hard to describe symptoms. Working on it. Probably one fix will do these both

Darkren committed:
Fix keep-alive and close packets initial send

Darkren committed:
Update vendor

Darkren committed:
Fix the issue when establishing the already existing route. Probably

Darkren committed:
Update vendor

Darkren commented:
Removed a couple of tests from checklist since they're not actual anymore

Darkren committed:
Change `io.EOF` to be `io.ErrClosedPipe` when RG is closed

Darkren committed:
Update vendor

Darkren committed:
Update vendor

Darkren committed:
Fix skysocks client to handle remote visor restart properly

Now we check if the yamux session is alive in different situations:
- stream accept failed
- session is closed after `io.Copy` returns
- session is checked with a time interval equal to half of the default route timeout

If session is broken, underlying app conn gets reestablished with the help of `netutil.Retrier`,
therefore route gets re-established (to be precise, the new route appears instead of the old one)

Darkren committed:
Update vendor

Darkren committed:
Refactor skysocks-client to preserve `skysocks.Client`'s original interface

Darkren committed:
Remove `github.com/pkg/errors` reference

Darkren committed:
Remove unused field from route group

Darkren commented:
> Would `RuleLastAccessed` be a better term than `RuleActivity`?

@evanlinjin sorry, can't find it. probably it was there before and now it's gone. if it's still actual, please let me know

Darkren commented:
1. We should discuss default keep alive timeout, we agreed on lowering them but we'll need to decide on the actual values. Also we should discuss the keep-alive packets interval
2. There're a lot of debug logs here which must be removed. But it's highly related to #171 which needs to be done first. I suggest that we finish #171 , remove debug logs and then merge this one

Darkren committed:
Update vendor

Darkren committed:
Update vendor

evanlinjin commented:
@Darkren I think we can merge after we sort out the logging situation.

nkryuchkov reviewed (approved):
Great job!

Darkren committed:
Refactor debug logs

Darkren committed:
Merge branch 'milestone2' of github.com:SkycoinProject/skywire-mainnet into fix/rule-expiration

Darkren committed:
Reduce keep-alive interval and route timeout
